rpc/src/client/subscription.rs

```diff
@@ -108,16 +109,37 @@ impl Subscription {
     }
 }
 
-/// A message sent to the subscription driver to terminate the subscription
-/// with the given parameters.
+/// A command that can be sent to the subscription driver.
 ///
-/// We expect the driver to use the `result_tx` channel to communicate the
-/// result of the termination request to the original caller.
+/// It is assumed that all [`SubscriptionClient`] implementations will follow a
+/// handle/driver concurrency model, where the client itself will just be a
+/// handle to a driver that runs in a separate coroutine.
+///
+/// [`SubscriptionClient`]: trait.SubscriptionClient.html
 #[derive(Debug, Clone)]
-pub struct TerminateSubscription {
-    pub id: SubscriptionId,
-    pub query: String,
-    pub result_tx: ChannelTx<Result<()>>,
+pub enum SubscriptionDriverCmd {
+    /// Initiate a new subscription.
+    Subscribe {
+        /// The desired ID for the new subscription.
+        id: SubscriptionId,
+        /// The query for which to initiate the subscription.
+        query: String,
+        /// Where to send events received for this subscription.
+        event_tx: ChannelTx<Result<Event>>,
+        /// Where to send the result of this subscription command.
+        result_tx: ChannelTx<Result<()>>,
+    },
+    /// Terminate an existing subscription.
+    Unsubscribe {
+        /// The ID of the subscription to terminate.
+        id: SubscriptionId,
+        /// The query associated with the subscription we want to terminate.
+        query: String,
```

Comment:
A bit weird that we'd need the query in here, you'd think the id is enough to unsubscribe, but I guess given the query->(id->sub) map in the router its easier to lookup the sub if we have both.

Reply:
Yeah exactly. Either approach (supplying the query, or only supplying the ID) is an optimization of some kind, and given that we don't expect super high rates of unsubscribe requests, neither approach should have any meaningful impact.

This approach of supplying the query (which we have stored in the `Subscription` anyways) does, however, allow me to avoid writing another little helper method in the `SubscriptionRouter` to look up a query by its subscription ID.